scalableminds/webknossos
# Auto-switch to a mapping/agglomerate if precomputed mesh uses that

I wrapped the two menu items which can be used to load a precomputed mesh with an HOC that asks the user to activate the corresponding mapping for the currently active mesh file before loading the mesh. The mapping activation is then triggered automatically and awaited, before the mesh loading is triggered.
The backend was adapted to send the designated mapping name for each mesh file if available or `null` otherwise.

Additionally, I performed some refactoring by moving mapping related code to a saga and extending the `isMappingEnabled` status to an enum with 3 states: DISABLED, ACTIVATING, ENABLED allowing easier selection of the layer relevant for json mappings and allowing to find out when a mapping activation has actually finished (with data reloading for agglomerate mappings).

### TODO:
- [x] ~~Confirm disappears in segment list if the segment list is rerendered which happens while hovering (not sure how to elegantly solve this)~~ No longer happens for me, not sure why :shrug: 
- [x] Test with a mesh file that has no mapping_name

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Open a dataset with a mesh file that was computed for a specific mapping and contains that information in the hdf5 metadata. For example this one:
[meshfile.zip](https://github.com/scalableminds/webknossos/files/7603183/meshfile.zip)
- Right click in a viewport and select "Load Mesh (precomputed)". You should be asked to confirm that the corresponding mapping will be activated. If confirmed, the mapping should be activated and the correct mesh should be loaded. If denied, nothing should happen.
  - Do the same from the segment list.
- Merger mode should still work.
- Manual (json) mapping activation should still work.

### Issues:
- fixes #5563
- contributes to https://github.com/scalableminds/webknossos/issues/5695#:~:text=mapping%20setup%20is,work%20right%20now.

------
(Please delete unneded items, merge only when none are left open)
